### PR TITLE
Support for OTP Flash as trust anchor for keystore

### PR DESCRIPTION
**Pull request by danielinux**
Goals:

 - [x] HAL for STM32H7
 - [x]  HAL for STM32H5
 - [X] Build option + keystore module for OTP flash
 - [x] `keygen` tool generates a one-time provisioning mechanism
 - [x] examples, .config with their build tests, documentation
 
(re)based on PR #400

### CODE REVIEW

File: tools/keytools/keygen.c [hunk 1]
```diff
@@ -200,7 +200,7 @@ const char Keystore_API[] =
                 "uint32_t keystore_get_mask(int id)\n"
                 "{\n"
                 "    if (id >= keystore_num_pubkeys())\n"
-                "        return -1;\n"
+                "        return 0;\n"
```
Comment (danielinux) [inline tools/keytools/keygen.c:203]: This looked like a bug in the generated software keystore, a mask with all "1" being returned for an invalid slot. Spotted while writing the API for the OTP trust anchor.

**Comment by danielinux** [comment]
Tested provisioning on a new STM32H563. OTP keystore working as expected. Ready for review.

File: tools/keytools/otp/Makefile [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+CC=$(CROSS_COMPILE)gcc
+OBJCOPY?=$(CROSS_COMPILE)objcopy
+SIZE?=$(CROSS_COMPILE)size
+
+
+otp-keystore-primer.bin: otp-keystore-primer.elf
+	@$(OBJCOPY) -O binary $(^) $(@)
+
+otp-keystore-primer.elf: $(PRI_KS_OBJS)
+	@$(CC) -o otp-keystore-primer.elf $(LDFLAGS) $(CFLAGS) $(PRI_KS_OBJS)
```
Comment (dgarske) [inline tools/keytools/otp/Makefile:36]: Might consider the V=1 logic and `$(Q)` logic.

Got a warning about missing entry symbol `_start`. Still built fine, but is the default entry at 0x8000 really okay?

```
% make clean && make
arm-none-eabi-gcc -D"__WOLFBOOT" -Werror -Wextra -Wno-array-bounds -mthumb -mlittle-endian -mthumb-interwork -DARCH_ARM -Ihal -mcpu=cortex-m33 -DCORTEX_M33 -mcmse -DWOLFCRYPT_SECURE_MODE -DWOLFSSL_SP_ASM -DWOLFSSL_SP_ARM_CORTEX_M_ASM -DTZEN -DARCH_FLASH_OFFSET=0x08000000 -DWOLFBOOT_DUALBOOT -DWOLFBOOT_ORIGIN=0x0C000000 -DBOOTLOADER_PARTITION_SIZE= -DWOLFBOOT_ARCH_ARM -DTARGET_stm32h5 -D"FLASH_OTP_KEYSTORE" -D"WOLFBOOT_SIGN_ECC256" -D"FILL_BYTE=0xFF" -D"DUALBANK_SWAP=1" -D"NVM_FLASH_WRITEONCE" -Os -D"WOLFBOOT_NO_MPU" -DWOLFBOOT_VERSION=1 -DSECURE_PKCS11 -DWOLFSSL_PKCS11_RW_TOKENS -DCK_CALLABLE="__attribute__((cmse_nonsecure_entry))" -Ilib/wolfPKCS11 -DWP11_HASH_PIN_COST=3 -D"WOLFBOOT_HASH_SHA256" -DIMAGE_HEADER_SIZE=256  "-Wstack-usage=16688" -I"." -I"include/" -I"lib/wolfssl" -Wno-array-bounds -D"WOLFSSL_USER_SETTINGS" -D"WOLFTPM_USER_SETTINGS" -D"PLATFORM_stm32h5" -Wall -Wextra -Wno-main -ffreestanding -Wno-unused -nostartfiles -ffunction-sections -fdata-sections -fomit-frame-pointer   -mthumb -mlittle-endian -mthumb-interwork -mcpu=cortex-m33 -Wl,-gc-sections -Wl,-Map=wolfboot.map -ffreestanding -nostartfiles  tools/keytools/otp/otp-keystore-primer.c   -o tools/keytools/otp/otp-keystore-primer
/usr/local/Cellar/gcc-arm-none-eabi/20200630/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld: warning: cannot find entry symbol _start; defaulting to 0000000000008000
	[CC-ARM] hal/stm32h5.o
```

File: include/wolfboot/wolfboot.h [hunk 3]
```diff
@@ -175,6 +175,10 @@ extern "C" {
     #endif
 #endif
 
+#endif
+
+#if defined __WOLFBOOT || defined __FLASH_OTP_PRIMER
```
Comment (dgarske) [inline include/wolfboot/wolfboot.h:180]: Maybe use `defined(__WOLFBOOT) || ` with parenthesis for consistency?

File: include/otp_keystore.h [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    #include "hal/stm32h5.h"
+#else
+    #error "Unsupported target for OTP"
+#endif
+
+#include "keystore.h"
+
+#define OTP_HDR_SIZE 16
+
+struct __attribute__((packed)) wolfBoot_otp_hdr {
```
Comment (dgarske) [inline include/otp_keystore.h:45]: Eventually I'd like to see our use of PACKED come from a macro, so it's not just GCC.

```
#if (defined(__IAR_SYSTEMS_ICC__) && (__IAR_SYSTEMS_ICC__ > 8)) || \
        defined(__GNUC__)
    #define WOLFBOOT_PACK __attribute__ ((packed))
#else
    #define WOLFBOOT_PACK
#endif
```

**Review by dgarske — CHANGES_REQUESTED** [review]
Builds good. Changes look good. Curious your thoughts on how best to test.

File: tools/keytools/otp/Makefile (hunk 2 above)
Comment (danielinux) [reply tools/keytools/otp/Makefile]: This was an actual issue with otp-keystore-primer build, not selecting the right linker script

**Review by dgarske — APPROVED** [review]

**Pull request merged**